Public/Src/Engine/Dll/FrontEndEngineImplementation.cs

```diff
@@ -274,13 +274,14 @@ public override async Task<ContentHash> GetFileContentHashAsync(string path, boo
 
                 VersionedFileIdentityAndContentInfo? maybeKnownIdentityAndHash = fileContentTable.TryGetKnownContentHash(fs);
 
-                if (maybeKnownIdentityAndHash.HasValue)
+                if (maybeKnownIdentityAndHash.HasValue && 
+                    (hashType == HashType.Unknown || maybeKnownIdentityAndHash.Value.FileContentInfo.Hash.HashType == hashType))

```

Comment:
Perhaps in `FileContentInfo`, you add the following method:

```
bool MatchHashType(HashType hashType) => hashType == HashType.Unknown /* unknown matches everything */ || HashType == hashType;
```

I believe that method can be useful when needed in other places.